### DISCUSSION TITLE
Throwing errors out of generated C code?

### DISCUSSION THREAD

**Discussion by amyworrall** [ricomariani/CG-SQL-author#83] [Q&A] [answered]
Sorry if my title doesn't make sense!

My problem: I've put a bunch of my business logic in CQL, all called from my app via the generated C code (with my own Swift codegen wrapper around it). I'm trying to decide what to do when operations fail for a business logic reason (as opposed to an infra reason) -- think, a field fails validation so my logic wants to refuse to save it to the DB.

I was wondering about using throw/catch, but have a few questions:

* It appears that CQL's throw, if not caught in some other CQL code, will make the result of the query be either the most recent SQLite error, or a generic `SQLITE_ERROR` if there isn't one. Then the result will be returned by the C function that calls my sproc. Is that correct?
* Is there any way to pass additional data out of CQL and into my C code via the throw/catch mechanism? Maybe a custom app-specific result code? 
* Is throw/catch the recommended way to solve my problem, or is there a better pattern I could consider? (I could use out parameters, or a custom resultset… I don't like either of those approaches that much because I'd have to do a bunch of work per sproc to set them up, rather than having one reusable mechanism.)

**Comment by ricomariani** [comment, marked as answer]
>It appears that CQL's throw, if not caught in some other CQL code, will make the result of the query be either the most recent SQLite error, or a generic SQLITE_ERROR if there isn't one. Then the result will be returned by the C function that calls my sproc. Is that correct?

Yes that is exactly right.  The code doesn't have to be a sqlite error code but it usually is.  If you avoid those codes you will see your own.  

>Is there any way to pass additional data out of CQL and into my C code via the throw/catch mechanism? Maybe a custom app-specific result code?

You can use your own result codes and your own helper function similar to `sqlite3_errmsg` 

>Is throw/catch the recommended way to solve my problem, or is there a better pattern I could consider? (I could use out parameters, or a custom resultset… I don't like either of those approaches that much because I'd have to do a bunch of work per sproc to set them up, rather than having one reusable mechanism.)

Try/catch is made for this.  Any other mechanism won't play nice with other pieces of code trying to clean up and rethrow.  You can do whatever you like if you stay in your own world but if you want to compose well use try/catch -- you use case is literally what it's for.  It also plays nicely with

```
proc foo()
begin
   proc savepoint
   begin
     -- your code that might fail
   end
end;
```
Which will save you a lot of grief

**Reply by amyworrall** [reply]
Brilliant, this confirms I'm on the right track.

So do I assign my own result code to `@RC` within the CQL code?

**Comment by ricomariani** [comment]
It's not quite that easy, @rc can't be directly mutated.  However, it's almost that easy.

```
-- add this to your sql
-- I am a procedure that uses the database and am allowed to throw
declare proc my_throw(x int!) using transation;

// link in this function somewhere
// I am the C implementation of a procedure that always throws
int my_throw(sqlite3 *db, int x) {
   return x;
}
```

Now you have a proc that throws what you want...

You can add other parameters such as strings or whatever, but remember you'll get a `cql_string_ref`, not a `char *` (there is a macro to convert).

If you're not sure how to code a more complicated C thing, you can just look at the compiler output, most things are dumb as rocks.

**Reply by ricomariani** [reply]
If you want to make an issue that says "there should be a builtin cql_throw" to throw the value of my choice I would not object and I could put such a thing in cqlrt_common.c in about 10s.

**Reply by ricomariani** [reply]
https://github.com/ricomariani/CG-SQL-author/commit/ec73ec83933fa1845760820b8f82c572030b18ae
